File: algorithms/79_word_search.py
class Solution(object):
    def exist(self, board, word):
        ROWS, COLS = len(board), len(board[0])
        path = set()
        def dfs(r, c, i):
            print(r, c)
            if i == len(word):
                return True
            if r>= ROWS or c >= COLS or word[i] != board[r][c] or (r, c) in path or r < 0 or c < 0:
                    return False
            path.add((r, c))
            res = dfs(r+1, c, i+1) or dfs(r, c+1, i+1)or dfs(r-1, c, i+1) or dfs(r, c-1, i+1)
            path.remove((r, c))
            return res
        for i in range(0, ROWS):
            for j in range(0, COLS):
                if dfs(i, j, 0):
                    return True
        return False
        

File: algorithms/79_word_search.java
class Solution {
    public Boolean dfs(int r, int c, int i, String word, char[][] board, HashSet<List<Integer>> set, int rows, int cols){
     if (i == word.length()){
         return true;
     }
     if(r >= rows || c >= cols|| r < 0 || c<0)  return false;
     if(word.charAt(i) != board[r][c]||
       set.contains(Arrays.asList(r, c)) ){
    
  
         return false;
     }
     set.add(Arrays.asList(r, c));
     Boolean res = (this.dfs(r+1, c, i+1, word, board, set, rows, cols)||
         this.dfs(r, c+1, i+1, word, board, set, rows, cols) ||
         this.dfs(r-1, c, i+1, word, board, set, rows, cols) ||
         this.dfs(r, c-1, i+1, word, board, set, rows, cols));
     set.remove(Arrays.asList(r, c));
     return res;
 }
 public boolean exist(char[][] board, String word) {
     final int ROWS = board.length;
     final int COLS = board[0].length;
     HashSet<List<Integer>> set = new HashSet<>();
     for (int i = 0; i < ROWS; i++){
         for (int j = 0; j < COLS; j++){
           
             if (board[i][j] == word.charAt(0)
                 && dfs(i, j, 0, word,board, set, ROWS, COLS)){
                 return true;
             }
         }
     }
     return false;
 }
 

 

 
}